db/query_context.hh

```diff
@@ -55,10 +55,18 @@ struct query_context {
                 // let the `storage_proxy` time out the query down the call chain
                 db::timeout_clock::duration::zero();
 
-        return do_with(timeout_config{d, d, d, d, d, d, d}, [this, req = std::move(req), &args...] (auto& tcfg) {
+        struct timeout_context {
+            std::unique_ptr<service::client_state> client_state;
```

Comment:
The context instance gets moved, and one of its attributes (`query_state`) takes reference to another one of its attributes (`client_state`). In order for this reference to remain valid, I have to ensure that `client_state` has a stable address - and the simplest solution is to put it on the heap. Not very nice from the performance perspective, but I also don't expect it to show on perf as top offender anywhere.

Reply:
This "not the top offender" ensures we can add allocations everywhere, since none of them will be a top offender.

But the untyped_result_set variants and not used in the data plane, so okay.

Reply:
@avikivity but note that this `execute_cql_with_timeout` helper function was introduced for paxos - which means that these extra allocation happens on LWT path. My gut tells me that it's still worth it to make internal calls use their own client state and that this allocation will not make a noticeable impact, but I haven't measured the actual overhead.
This allocation of a new client state was avoided so far, because paxos effectively used a global client state with specific timeout config, but it's no longer possible after timeout configuration is moved to the client state itself.

Reply:
On the other hand, if we really care about optimizing this path, raw CQL strings should just be dropped altogether, the whole cql layer should be bypassed for paxos and the request should go straight to storage proxy, like we do in alternator. So this single allocation is not the root cause of potential slowness.

Reply:
> one of its attributes (query_state) takes reference to another one of its attributes (client_state)

Please explain that in client_state's blurb.